turtleFSI/modules/newtonsolver.py

```diff
@@ -64,20 +64,26 @@ def newtonsolver(F, J_nonlinear, A_pre, A, b, bcs, lmbda, recompute, recompute_t
         if recompute_for_timestep or recompute_frequency or recompute_residual or recompute_initialize:
             if MPI.rank(MPI.comm_world) == 0 and verbose:
                 print("Compute Jacobian matrix")
+            # Assemble non-linear part of Jacobian, keep sparsity pattern (keep_diagonal=True)
             assemble(J_nonlinear, tensor=A,
                          form_compiler_parameters=compiler_parameters,
                          keep_diagonal=True)
+            # Add non-linear and linear part of Jacobian
             A.axpy(1.0, A_pre, True)
+            # Insert ones on diagonal to make sure the matrix is non-singular (related to solid pressure being zero)
             A.ident_zeros()
             [bc.apply(A) for bc in bcs]
 
-        # Compute right hand side
-        b = assemble(-F, tensor=b)
+        # Aseemble right hand side vector
+        b = assemble(-F)
```

Comment:
Why do we not use the `tensor=b` any longer?

Reply:
I thought the same argument for `A = assemble(J_nonlinear, tensor=A)` would apply to `b` and thus deleted. Do we need to specify `tensor=b` even though we assign it to `b` anyway?

Reply:
```
assemble(-F, tensor=b)
```
produced the following error when applying bcs 
```
 File "/Users/keiyamamoto/Documents/turtleFSI/turtleFSI/modules/newtonsolver.py", line 81, in <listcomp>
    [bc.apply(b, dvp_["n"].vector()) for bc in bcs]
TypeError: apply(): incompatible function arguments. The following argument types are supported:
    1. (self: dolfin.cpp.fem.DirichletBC, arg0: dolfin::GenericVector) -> None
    2. (self: dolfin.cpp.fem.DirichletBC, arg0: dolfin::GenericMatrix) -> None
    3. (self: dolfin.cpp.fem.DirichletBC, arg0: dolfin::GenericMatrix, arg1: dolfin::GenericVector) -> None
    4. (self: dolfin.cpp.fem.DirichletBC, arg0: dolfin::GenericVector, arg1: dolfin::GenericVector) -> None
    5. (self: dolfin.cpp.fem.DirichletBC, arg0: dolfin::GenericMatrix, arg1: dolfin::GenericVector, arg2: dolfin::GenericVector) -> None

Invoked with: <dolfin.fem.dirichletbc.DirichletBC object at 0x180732cf0>, None, <dolfin.cpp.la.PETScVector object at 0x178d40220>
```
Is it because `b` is defined as `None`?
https://github.com/KVSlab/turtleFSI/blob/d149375c2d89f17e5662f352e4547ffd1a471750/turtleFSI/modules/newtonsolver.py#L28

Reply:
Then we should have tests for all assembly steps like:
```python
if b is None:
    b = assemble(-F)
else:
    assemble(b, tensor=b)
```
and similarly wherever we do `assemble(..,tensor=A)`
so that we can cache things if they have been already created.

Reply:
At least add a note that we assume that `A` is not `None` when it is input to the `newtonsolver` (because someone might stumble over this in the future and remove `A` from the pre-creation.

Reply:
added a note on that!